### ISSUE TITLE
Visibility of brotli APIs on Unix

### ISSUE THREAD

**Issue by kasper3** [google/brotli#687]
For Unix, we had to apply this patch to set the visibility of BROTLI_ENC_API dotnet/corefx@0c770ca72daf50ad12703b175dcb83e306664ba6.  This happens when downstream is compiling with `-fvisibility=hidden`. We have out-of-the-tree cmake file that includes Brotli's sources.

@eustas, could this patch be also considered for Brotli vNext? AFAICT, there are no ramifications of this change, when the consumer doesn't have `-fvisibility` set.

**Comment by eustas** [comment]
Sure. Perhaps with a little bit different enclosure, i.e.
```
if shared
  if win32
    XXX
  else
    YYY
else
  ZZZ
```

**Comment by kasper3** [comment]
That would certainly improve readability. 👍

**Comment by kasper3** [comment]
I think it would look like:

```h
#if defined(BROTLI_SHARED_COMPILATION)
  #if defined(_WIN32)
    #if defined(BROTLICOMMON_SHARED_COMPILATION)
      #define BROTLI_COMMON_API __declspec(dllexport)
    #else /* BROTLICOMMON_SHARED_COMPILATION */
      #define BROTLI_COMMON_API __declspec(dllimport)
    #endif /* BROTLICOMMON_SHARED_COMPILATION */
    #if defined(BROTLIDEC_SHARED_COMPILATION)
      #define BROTLI_DEC_API __declspec(dllexport)
    #else /* BROTLIDEC_SHARED_COMPILATION */
      #define BROTLI_DEC_API __declspec(dllimport)
    #endif /* BROTLIDEC_SHARED_COMPILATION */
    #if defined(BROTLIENC_SHARED_COMPILATION)
      #define BROTLI_ENC_API __declspec(dllexport)
    #else /* BROTLIENC_SHARED_COMPILATION */
      #define BROTLI_ENC_API __declspec(dllimport)
    #endif /* BROTLIENC_SHARED_COMPILATION */
  #else /* _WIN32 */
    #define BROTLI_COMMON_API __attribute__ ((__visibility__ ("default")))
    #define BROTLI_DEC_API __attribute__ ((__visibility__ ("default")))
    #define BROTLI_ENC_API __attribute__ ((__visibility__ ("default")))
  #endif /* _WIN32 */
#else /* BROTLI_SHARED_COMPILATION */
  #define BROTLI_COMMON_API
  #define BROTLI_DEC_API
  #define BROTLI_ENC_API
#endif /* BROTLI_SHARED_COMPILATION */
```

**Comment by eustas** [comment]
Should be fixed with #688

**Comment by kasper3** [comment]
Thanks! Waiting for the next release 👏
